rainforest/mobile_upload.go

```diff
@@ -0,0 +1,212 @@
+package rainforest
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime/multipart"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+)
+
+// RFPresignedPostData represents the url and required fields we must POST to AWS
+// in order to upload
+type RFPresignedPostData struct {
+	URL            string            `json:"url"`
+	RequiredFields map[string]string `json:"url_fields"`
+	RainforestURL  string            `json:"rainforest_url"`
+}
+
+// GetPresignedPOST requests the presigned POST data from Rainforest so that we can upload the mobile
+// app to S3
+func (c *Client) GetPresignedPOST(fileExt string, siteID int, environmentID int) (*RFPresignedPostData, error) {
+	var data RFPresignedPostData
+	url := "uploads"
+	req, err := c.NewRequest("GET", url, nil)
+	q := req.URL.Query()
+	q.Add("app_site_id", strconv.Itoa(siteID))
+	q.Add("environment_id", strconv.Itoa(environmentID))
+	q.Add("extension", fileExt)
+	req.URL.RawQuery = q.Encode()
+
+	_, err = c.Do(req, &data)
+
+	return &data, err
+}
+
+// UploadToS3 creates a http.Request containing the required body for
+// uploading a file to AWS given the values stored in the receiving awsFileInfo struct.
+func (c *Client) UploadToS3(postData *RFPresignedPostData, filePath string) error {
+	var req *http.Request
+	fileName := filepath.Base(filePath)
+
+	f, err := os.Open(filePath)
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		return err
+	}
+	fileSize := int64(fi.Size())
+
+	contentLength := postData.emptyMultipartSize("file", fileName) + fileSize
+
+	readBody, writeBody := io.Pipe()
+	defer readBody.Close()
+
+	writer := multipart.NewWriter(writeBody)
+
+	// Do the writes async streamed
+	errChan := make(chan error, 1)
+	go func() {
+		defer writeBody.Close()
+
+		// Add the required fields from S3
+		addFields(postData.RequiredFields, writer)
+
+		part, err := writer.CreateFormFile("file", fileName)
+		if err != nil {
+			errChan <- err
+			return
+		}
+		if _, err := io.CopyN(part, f, fileSize); err != nil {
+			errChan <- err
+			return
+		}
+		errChan <- writer.Close()
+	}()
+
+	// Create the Request
+	url := postData.URL
+	req, err = http.NewRequest("POST", url, readBody)
+	if err != nil {
+		<-errChan
+		return err
+	}
+	req.Header.Set("Content-Type", writer.FormDataContentType())
+	req.ContentLength = contentLength
+
+	// Perform the upload
+	client := &http.Client{}
+	resp, err := client.Do(req)
```

Comment:
The other goroutine needs to complete writing before we send this request, right? I think we have to way on `<-errChan` before this.

Reply:
That `Do` is what uses the pipe to stream the request and not load it into memory.